README.pod
=pod

=encoding utf-8

=head1 NAME

CPAN::Common::Index - Common library for searching CPAN modules, authors and distributions

=head1 VERSION

version 0.004

=head1 SYNOPSIS

    use CPAN::Common::Index::Mux::Ordered;
    use Data::Dumper;

    $index = CPAN::Common::Index::Mux::Ordered->assemble(
        MetaDB => {},
        Mirror => { mirror => "http://cpan.cpantesters.org" },
    );

    $result = $index->search_packages( { package => "Moose" } );

    print Dumper($result);

    # {
    #   package => 'MOOSE',
    #   version => '2.0802',
    #   uri     => "cpan:///distfile/ETHER/Moose-2.0802.tar.gz"
    # }

=head1 DESCRIPTION

This module provides a common library for working with a variety of CPAN index
services.  It is intentionally minimalist, trying to use as few non-core
modules as possible.

The C<CPAN::Common::Index> module is an abstract base class that defines a
common API.  Individual backends deliver the API for a particular index.

As shown in the SYNOPSIS, one interesting application is multiplexing -- using
different index backends, querying each in turn, and returning the first
result.

=head1 METHODS

=head2 search_packages (ABSTRACT)

    $result = $index->search_packages( { package => "Moose" });
    @result = $index->search_packages( \%advanced_query );

Searches the index for a package such as listed in the CPAN
F<02packages.details.txt> file.  The query must be provided as a hash
reference.  Valid keys are

=over 4

=item *

package -- a string, regular expression or code reference

=item *

version -- a version number or code reference

=item *

dist -- a string, regular expression or code reference

=back

If the query term is a string or version number, the query will be for an exact
match.  If a code reference, the code will be called with the value of the
field for each potential match.  It should return true if it matches.

Not all backends will implement support for all fields or all types of queries.
If it does not implement either, it should "decline" the query with an empty
return.

The return should be context aware, returning either a
single result or a list of results.

The result must be formed as follows:

    {
      package => 'MOOSE',
      version => '2.0802',
      uri     => "cpan:///distfile/ETHER/Moose-2.0802.tar.gz"
    }

The C<uri> field should a valid URI.  It may be a L<URI::cpan> or any other
URI.  (It is up to a client to do something useful with any given URI scheme.)

=head2 search_authors (ABSTRACT)

    $result = $index->search_authors( { id => "DAGOLDEN" });
    @result = $index->search_authors( \%advanced_query );

Searches the index for author data such as from the CPAN F<01mailrc.txt> file.
The query must be provided as a hash reference.  Valid keys are

=over 4

=item *

id -- a string, regular expression or code reference

=item *

fullname -- a string, regular expression or code reference

=item *

email -- a string, regular expression or code reference

=back

If the query term is a string, the query will be for an exact match.  If a code
reference, the code will be called with the value of the field for each
potential match.  It should return true if it matches.

Not all backends will implement support for all fields or all types of queries.
If it does not implement either, it should "decline" the query with an empty
return.

The return should be context aware, returning either a single result or a list
of results.

The result must be formed as follows:

    {
        id       => 'DAGOLDEN',
        fullname => 'David Golden',
        email    => 'dagolden@cpan.org',
    }

The C<email> field may not reflect an actual email address.  The 01mailrc file
on CPAN often shows "CENSORED" when email addresses are concealed.

=head2 index_age

    $epoch = $index->index_age;

Returns the modification time of the index in epoch seconds.  This may not make sense
for some backends.  By default it returns the current time.

=head2 refresh_index

    $index->refresh_index;

This ensures the index source is up to date.  For example, a remote
mirror file would be re-downloaded.  By default, it does nothing.

=head2 attributes

Return attributes and default values as a hash reference.  By default
returns an empty hash reference.

=head2 validate_attributes

    $self->validate_attributes;

This is called by the constructor to validate any arguments.  Subclasses
should override the default one to perform validation.  It should not be
called by application code.  By default, it does nothing.

=for Pod::Coverage method_names_here

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/dagolden/cpan-common-index/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/dagolden/cpan-common-index>

  git clone git://github.com/dagolden/cpan-common-index.git

=head1 AUTHOR

David Golden <dagolden@cpan.org>

=head1 CONTRIBUTOR

Tatsuhiko Miyagawa <miyagawa@bulknews.net>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2013 by David Golden.

This is free software, licensed under:

  The Apache License, Version 2.0, January 2004



Changes
Revision history for CPAN-Common-Index

{{$NEXT}}

0.010     2017-07-26 10:36:48-04:00 America/New_York

    - No changes from 0.009

0.009     2017-06-08 14:03:34-04:00 America/New_York (TRIAL RELEASE)

    [CHANGED]

    - Make IO::Uncompress::Gunzip optional for fat-pack-ability.

0.007     2017-05-11 13:07:37-04:00 America/New_York

    [CHANGED]

    - Removed dependency on Path::Tiny; uses core modules directly.

0.006     2015-04-29 16:53:04-04:00 America/New_York

    [FIXED]

    - Fixed package name search when there are packages that
      differ only in case (e.g. "If" vs "if")

0.005     2014-09-23 21:26:13-04:00 America/New_York

    [FIXED]

    - Fixed tests for Class::Tiny 1.000 changes

0.004     2014-03-21 06:56:46-04:00 America/New_York

    [FIXED]

    - Fixed bug removing tempfiles on Windows (Kenichi Ishigaki)

    [CHANGED]

    - Modernized repo meta and support files

0.003     2013-08-14 16:18:00 America/New_York

    [FIXED]

    - Package searches are now correctly case sensitive

    - Fixed incorrect synopses

0.002     2013-06-27 22:10:59 America/New_York

    [FIXED]

    - Bumped Search::Dict prereq

0.001     2013-06-21 18:43:14 America/New_York

    - First release



t/lib/CommonTests.pm
use strict;
use warnings;

package CommonTests;

use Exporter;
use Test::More;

our @ISA    = qw/Exporter/;
our @EXPORT = qw(
  test_find_package
  test_search_package
  test_find_author
  test_search_author
);

sub test_find_package {
    my $index = shift;

    my @cases = (
        {
            label  => 'query on package File::Marker',
            query  => { package => "File::Marker" },
            result => {
                package => 'File::Marker',
                version => '0.13',
                uri     => 'cpan:///distfile/DAGOLDEN/File-Marker-0.13.tar.gz',
            },
        },
        {
            label  => 'query on package dist::zilla',
            query  => { package => "dist::zilla" },
            result => undef,                         # should not match case insensitively
        },
        {
            label  => 'query on package without version ',
            query  => { package => "Moo::Role" },
            result => {
                package => 'Moo::Role',
                version => 'undef',
                uri     => 'cpan:///distfile/MSTROUT/Moo-1.001000.tar.gz',
            },
        },
        {
            label  => 'query on package in a perl distribution',
            query  => { package => 'attributes', },
            result => {
                package => 'attributes',
                version => '0.2',
                uri     => 'cpan:///distfile/FLORA/perl-5.17.4.tar.bz2',
            },
        },
    );

    for my $c (@cases) {
        my $got = $index->search_packages( $c->{query} );
        is_deeply( $got, $c->{result}, $c->{label} ) or diag explain $got;
    }
}

sub test_search_package {
    my $index = shift;

    my @cases = (
        {
            label  => 'query on package',
            query  => { package => qr/e::Marker$/, },
            result => [
                {
                    package => 'File::Marker',
                    version => '0.13',
                    uri     => 'cpan:///distfile/DAGOLDEN/File-Marker-0.13.tar.gz',
                }
            ],
        },
        {
            label => 'query on package and version',
            query => {
                package => qr/Marker$/,
                version => 0.13,
            },
            result => [
                {
                    package => 'File::Marker',
                    version => '0.13',
                    uri     => 'cpan:///distfile/DAGOLDEN/File-Marker-0.13.tar.gz',
                }
            ],
        },
        {
            label  => 'query on dist',
            query  => { dist => qr/1\.4404\.tar\.gz$/, },
            result => [
                {
                    'package' => 'Parse::CPAN::Meta',
                    'uri'     => 'cpan:///distfile/DAGOLDEN/Parse-CPAN-Meta-1.4404.tar.gz',
                    'version' => '1.4404'
                }
            ],
        },
    );

    for my $c (@cases) {
        my @got = $index->search_packages( $c->{query} );
        is_deeply( \@got, $c->{result}, $c->{label} ) or diag explain \@got;
    }
}

sub test_find_author {
    my $index = shift;

    my @cases = (
        {
            label  => 'query on DAGOLDEN',
            query  => { id => 'DAGOLDEN' },
            result => {
                id       => 'DAGOLDEN',
                fullname => 'David Golden',
                email    => 'dagolden@cpan.org',
            },
        },
        {
            label  => 'query on aashu (with CENSORED email)',
            query  => { id => 'aashu' },
            result => {
                id       => 'AASHU',
                fullname => 'Ashutosh Sharma',
                email    => 'CENSORED',
            },
        },
    );

    for my $c (@cases) {
        my $got = $index->search_authors( $c->{query} );
        is_deeply( $got, $c->{result}, $c->{label} ) or diag explain $got;
    }
}

sub test_search_author {
    my $index = shift;

    my @cases = (
        {
            label  => 'query id on qr/DAGOLD/',
            query  => { id => qr/DAGOLD/, },
            result => [
                {
                    id       => 'DAGOLDEN',
                    fullname => 'David Golden',
                    email    => 'dagolden@cpan.org',
                },
            ],
        },
        {
            label  => 'query email on qr/dagolden/',
            query  => { email => qr/dagolden/ },
            result => [
                {
                    id       => 'DAGOLDEN',
                    fullname => 'David Golden',
                    email    => 'dagolden@cpan.org',
                },
            ],
        },
        {
            label  => 'query fullname on qr/Golden$/',
            query  => { fullname => qr/Golden$/ },
            result => [
                {
                    id       => 'DAGOLDEN',
                    fullname => 'David Golden',
                    email    => 'dagolden@cpan.org',
                },
            ],
        },
    );

    for my $c (@cases) {
        my @got = $index->search_authors( $c->{query} );
        is_deeply( \@got, $c->{result}, $c->{label} ) or diag explain \@got;
    }
}

1;



lib/CPAN/Common/Index.pm
use 5.008001;
use strict;
use warnings;

package CPAN::Common::Index;
# ABSTRACT: Common library for searching CPAN modules, authors and distributions

our $VERSION = '0.011';

use Carp ();

use Class::Tiny;

#--------------------------------------------------------------------------#
# Document abstract methods
#--------------------------------------------------------------------------#

=method search_packages (ABSTRACT)

    $result = $index->search_packages( { package => "Moose" });
    @result = $index->search_packages( \%advanced_query );

Searches the index for a package such as listed in the CPAN
F<02packages.details.txt> file.  The query must be provided as a hash
reference.  Valid keys are

=for :list
* package -- a string, regular expression or code reference
* version -- a version number or code reference
* dist -- a string, regular expression or code reference

If the query term is a string or version number, the query will be for an exact
match.  If a code reference, the code will be called with the value of the
field for each potential match.  It should return true if it matches.

Not all backends will implement support for all fields or all types of queries.
If it does not implement either, it should "decline" the query with an empty
return.

The return should be context aware, returning either a
single result or a list of results.

The result must be formed as follows:

    {
      package => 'MOOSE',
      version => '2.0802',
      uri     => "cpan:///distfile/ETHER/Moose-2.0802.tar.gz"
    }

The C<uri> field should be a valid URI.  It may be a L<URI::cpan> or any other
URI.  (It is up to a client to do something useful with any given URI scheme.)

=method search_authors (ABSTRACT)

    $result = $index->search_authors( { id => "DAGOLDEN" });
    @result = $index->search_authors( \%advanced_query );

Searches the index for author data such as from the CPAN F<01mailrc.txt> file.
The query must be provided as a hash reference.  Valid keys are

=for :list
* id -- a string, regular expression or code reference
* fullname -- a string, regular expression or code reference
* email -- a string, regular expression or code reference

If the query term is a string, the query will be for an exact match.  If a code
reference, the code will be called with the value of the field for each
potential match.  It should return true if it matches.

Not all backends will implement support for all fields or all types of queries.
If it does not implement either, it should "decline" the query with an empty
return.

The return should be context aware, returning either a single result or a list
of results.

The result must be formed as follows:

    {
        id       => 'DAGOLDEN',
        fullname => 'David Golden',
        email    => 'dagolden@cpan.org',
    }

The C<email> field may not reflect an actual email address.  The 01mailrc file
on CPAN often shows "CENSORED" when email addresses are concealed.

=cut

#--------------------------------------------------------------------------#
# stub methods
#--------------------------------------------------------------------------#

=method index_age

    $epoch = $index->index_age;

Returns the modification time of the index in epoch seconds.  This may not make sense
for some backends.  By default it returns the current time.

=cut

sub index_age { time }

=method refresh_index

    $index->refresh_index;

This ensures the index source is up to date.  For example, a remote
mirror file would be re-downloaded.  By default, it does nothing.

=cut

sub refresh_index { 1 }

=method attributes

Return attributes and default values as a hash reference.  By default
returns an empty hash reference.

=cut

sub attributes { {} }

=method validate_attributes

    $self->validate_attributes;

This is called by the constructor to validate any arguments.  Subclasses
should override the default one to perform validation.  It should not be
called by application code.  By default, it does nothing.

=cut

sub validate_attributes { 1 }

1;

=for Pod::Coverage method_names_here

=head1 SYNOPSIS

    use CPAN::Common::Index::Mux::Ordered;
    use Data::Dumper;

    $index = CPAN::Common::Index::Mux::Ordered->assemble(
        MetaDB => {},
        Mirror => { mirror => "http://cpan.cpantesters.org" },
    );

    $result = $index->search_packages( { package => "Moose" } );

    print Dumper($result);

    # {
    #   package => 'MOOSE',
    #   version => '2.0802',
    #   uri     => "cpan:///distfile/ETHER/Moose-2.0802.tar.gz"
    # }

=head1 DESCRIPTION

This module provides a common library for working with a variety of CPAN index
services.  It is intentionally minimalist, trying to use as few non-core
modules as possible.

The C<CPAN::Common::Index> module is an abstract base class that defines a
common API.  Individual backends deliver the API for a particular index.

As shown in the SYNOPSIS, one interesting application is multiplexing -- using
different index backends, querying each in turn, and returning the first
result.

=cut

# vim: ts=4 sts=4 sw=4 et:


lib/CPAN/Common/Index/Util.pm
use 5.008001;
use strict;
use warnings;

package CPAN::Common::Index::Util;
# ABSTRACT: Utility functions for CPAN::Common::Index

our $VERSION = '0.008';

sub gunzip {
    my ($src, $dest) = @_;
    if ( eval { require IO::Uncompress::Gunzip } ) {
        no warnings 'once';
        IO::Uncompress::Gunzip::gunzip($src, $dest)
          or die "gunzip failed: $IO::Uncompress::Gunzip::GunzipError\n";
        return 1;
    }
    else {
        require File::Which;
        require IPC::Run3;
        my $gzip = File::Which::which("gzip")
          or die "can't find IO::Uncompress::Gunzip nor gzip command\n";
        IPC::Run3::run3( [$gzip, '--decompress', '--stdout', $src], \undef, $dest, \my $err );
        return 1 if $? == 0;
        chomp $err;
        die $err ? $err : "gunzip failed: \$? = $?", "\n";
    }
}

1;


lib/CPAN/Common/Index/MetaDB.pm
use 5.008001;
use strict;
use warnings;

package CPAN::Common::Index::MetaDB;
# ABSTRACT: Search index via CPAN MetaDB

our $VERSION = '0.011';

use parent 'CPAN::Common::Index';

use Class::Tiny qw/uri/;

use Carp;
use CPAN::Meta::YAML;
use CPAN::Meta::Requirements;
use HTTP::Tiny;

=attr uri

A URI for the endpoint of a CPAN MetaDB server. The
default is L<http://cpanmetadb.plackperl.org/v1.0/>.

=cut

sub BUILD {
    my $self = shift;
    my $uri  = $self->uri;
    $uri = "http://cpanmetadb.plackperl.org/v1.0/"
      unless defined $uri;
    # ensure URI ends in '/'
    $uri =~ s{/?$}{/};
    $self->uri($uri);
    return;
}

sub search_packages {
    my ( $self, $args ) = @_;
    Carp::croak("Argument to search_packages must be hash reference")
      unless ref $args eq 'HASH';

    return
      unless exists $args->{package} && ref $args->{package} eq '';

    my $mod = $args->{package};

    if ($args->{version} || $args->{version_range}) {
        my $res = HTTP::Tiny->new->get( $self->uri . "history/$mod" );
        return unless $res->{success};

        my $range = defined $args->{version} ? "== $args->{version}" : $args->{version_range};
        my $reqs = CPAN::Meta::Requirements->from_string_hash({ $mod => $range });

        my @found;
        for my $line ( split /\r?\n/, $res->{content} ) {
            if ($line =~ /^$mod\s+(\S+)\s+(\S+)$/) {
                push @found, {
                    version => $1,
                    version_o => version::->parse($1),
                    distfile => $2,
                };
            }
        }

        my $match;
        for my $try (sort { $b->{version_o} <=> $a->{version_o} } @found) {
            if ($reqs->accepts_module($mod => $try->{version_o})) {
                $match = $try, last;
            }
        }

        if ($match) {
            my $file = $match->{distfile};
            $file =~ s{^./../}{}; # strip leading
            return {
                package => $mod,
                version => $match->{version},
                uri     => "cpan:///distfile/$file",
            };
        }
    } else {
        my $res = HTTP::Tiny->new->get( $self->uri . "package/$mod" );
        return unless $res->{success};

        if ( my $yaml = CPAN::Meta::YAML->read_string( $res->{content} ) ) {
            my $meta = $yaml->[0];
            if ( $meta && $meta->{distfile} ) {
                my $file = $meta->{distfile};
                $file =~ s{^./../}{}; # strip leading
                return {
                    package => $mod,
                    version => $meta->{version},
                    uri     => "cpan:///distfile/$file",
                };
            }
        }
    }

    return;
}

sub index_age { return time };    # pretend always current

sub search_authors { return };    # not supported

1;

=for Pod::Coverage attributes validate_attributes search_packages search_authors BUILD

=head1 SYNOPSIS

  use CPAN::Common::Index::MetaDB;

  $index = CPAN::Common::Index::MetaDB->new;

  $index->search_packages({ package => "Moose" });
  $index->search_packages({ package => "Moose", version_range => ">= 2.0" });

=head1 DESCRIPTION

This module implements a CPAN::Common::Index that searches for packages against
the same CPAN MetaDB API used by L<cpanminus>.

There is no support for advanced package queries or searching authors.  It just
takes a package name and returns the corresponding version and distribution.

=cut

# vim: ts=4 sts=4 sw=4 et:


lib/CPAN/Common/Index/Mirror.pm
use 5.008001;
use strict;
use warnings;

package CPAN::Common::Index::Mirror;
# ABSTRACT: Search index via CPAN mirror flatfiles

our $VERSION = '0.011';

use parent 'CPAN::Common::Index';

use Class::Tiny qw/cache mirror/;

use Carp;
use CPAN::DistnameInfo;
use File::Basename ();
use File::Fetch;
use File::Temp 0.19; # newdir
use Search::Dict 1.07;
use Tie::Handle::SkipHeader;
use URI;

our $HAS_IO_UNCOMPRESS_GUNZIP = eval { require IO::Uncompress::Gunzip };

=attr mirror

URI to a CPAN mirror.  Defaults to C<http://www.cpan.org/>.

=attr cache

Path to a local directory to store copies of the source indices.  Defaults to a
temporary directory if not specified.

=cut

sub BUILD {
    my $self = shift;

    # cache directory needs to exist
    my $cache = $self->cache;
    $cache = File::Temp->newdir
      unless defined $cache;
    if ( !-d $cache ) {
        Carp::croak("Cache directory '$cache' does not exist");
    }
    $self->cache($cache);

    # ensure mirror URL ends in '/'
    my $mirror = $self->mirror;
    $mirror = "http://www.cpan.org/"
      unless defined $mirror;
    $mirror =~ s{/?$}{/};
    $self->mirror($mirror);

    return;
}

my %INDICES = (
    mailrc   => 'authors/01mailrc.txt.gz',
    packages => 'modules/02packages.details.txt.gz',
);

# XXX refactor out from subs below
my %TEST_GENERATORS = (
    regexp_nocase => sub {
        my $arg = shift;
        my $re = ref $arg eq 'Regexp' ? $arg : qr/\A\Q$arg\E\z/i;
        return sub { $_[0] =~ $re };
    },
    regexp => sub {
        my $arg = shift;
        my $re = ref $arg eq 'Regexp' ? $arg : qr/\A\Q$arg\E\z/;
        return sub { $_[0] =~ $re };
    },
    version => sub {
        my $arg = shift;
        my $v   = version->parse($arg);
        return sub {
            eval { version->parse( $_[0] ) == $v };
        };
    },
);

my %QUERY_TYPES = (
    # package search
    package => 'regexp',
    version => 'version',
    dist    => 'regexp',

    # author search
    id       => 'regexp_nocase', # XXX need to add "alias " first
    fullname => 'regexp_nocase',
    email    => 'regexp_nocase',
);

sub cached_package {
    my ($self) = @_;
    my $package = File::Spec->catfile( $self->cache,
        File::Basename::basename( $INDICES{packages} ) );
    $package =~ s/\.gz$//;
    $self->refresh_index unless -r $package;
    return $package;
}

sub cached_mailrc {
    my ($self) = @_;
    my $mailrc =
      File::Spec->catfile( $self->cache, File::Basename::basename( $INDICES{mailrc} ) );
    $mailrc =~ s/\.gz$//;
    $self->refresh_index unless -r $mailrc;
    return $mailrc;
}

sub refresh_index {
    my ($self) = @_;
    for my $file ( values %INDICES ) {
        my $remote = URI->new_abs( $file, $self->mirror );
        $remote =~ s/\.gz$//
          unless $HAS_IO_UNCOMPRESS_GUNZIP;
        my $ff = File::Fetch->new( uri => $remote );
        my $where = $ff->fetch( to => $self->cache )
          or Carp::croak( $ff->error );
        if ($HAS_IO_UNCOMPRESS_GUNZIP) {
            ( my $uncompressed = $where ) =~ s/\.gz$//;
            no warnings 'once';
            IO::Uncompress::Gunzip::gunzip( $where, $uncompressed )
              or Carp::croak "gunzip failed: $IO::Uncompress::Gunzip::GunzipError\n";
        }
    }
    return 1;
}

# epoch secs
sub index_age {
    my ($self) = @_;
    my $package = $self->cached_package;
    return ( -r $package ? ( stat($package) )[9] : 0 ); # mtime if readable
}

sub search_packages {
    my ( $self, $args ) = @_;
    Carp::croak("Argument to search_packages must be hash reference")
      unless ref $args eq 'HASH';

    my $index_path = $self->cached_package;
    die "Can't read $index_path" unless -r $index_path;

    my $fh = IO::Handle->new;
    tie *$fh, 'Tie::Handle::SkipHeader', "<", $index_path
      or die "Can't tie $index_path: $!";

    # Convert scalars or regexps to subs
    my $rules;
    while ( my ( $k, $v ) = each %$args ) {
        $rules->{$k} = _rulify( $k, $v );
    }

    my @found;
    if ( $args->{package} and ref $args->{package} eq '' ) {
        # binary search 02packages on package
        my $pos = look $fh, $args->{package}, { xfrm => \&_xform_package, fold => 1 };
        return if $pos == -1;
        # loop over any case-insensitive matching lines
        LINE: while ( my $line = <$fh> ) {
            last unless $line =~ /\A\Q$args->{package}\E\s+/i;
            push @found, _match_package_line( $line, $rules );
        }
    }
    else {
        # iterate all lines looking for match
        LINE: while ( my $line = <$fh> ) {
            push @found, _match_package_line( $line, $rules );
        }
    }
    return wantarray ? @found : $found[0];
}

sub search_authors {
    my ( $self, $args ) = @_;
    Carp::croak("Argument to search_authors must be hash reference")
      unless ref $args eq 'HASH';

    my $index_path = $self->cached_mailrc;
    die "Can't read $index_path" unless -r $index_path;
    open my $fh, $index_path or die "Can't open $index_path: $!";

    # Convert scalars or regexps to subs
    my $rules;
    while ( my ( $k, $v ) = each %$args ) {
        $rules->{$k} = _rulify( $k, $v );
    }

    my @found;
    if ( $args->{id} and ref $args->{id} eq '' ) {
        # binary search mailrec on package
        my $pos = look $fh, $args->{id}, { xfrm => \&_xform_mailrc, fold => 1 };
        return if $pos == -1;
        my $line = <$fh>;
        push @found, _match_mailrc_line( $line, $rules );
    }
    else {
        # iterate all lines looking for match
        LINE: while ( my $line = <$fh> ) {
            push @found, _match_mailrc_line( $line, $rules );
        }
    }
    return wantarray ? @found : $found[0];
}

sub _rulify {
    my ( $key, $arg ) = @_;
    return $arg if ref($arg) eq 'CODE';
    return $TEST_GENERATORS{ $QUERY_TYPES{$key} }->($arg);
}

sub _xform_package {
    my @fields = split " ", $_[0], 2;
    return $fields[0];
}

sub _xform_mailrc {
    my @fields = split " ", $_[0], 3;
    return $fields[1];
}

sub _match_package_line {
    my ( $line, $rules ) = @_;
    return unless defined $line;
    my ( $mod, $version, $dist, $comment ) = split " ", $line, 4;
    if ( $rules->{package} ) {
        return unless $rules->{package}->($mod);
    }
    if ( $rules->{version} ) {
        return unless $rules->{version}->($version);
    }
    if ( $rules->{dist} ) {
        return unless $rules->{dist}->($dist);
    }
    $dist =~ s{\A./../}{};
    return {
        package => $mod,
        version => $version,
        uri     => "cpan:///distfile/$dist",
    };
}

sub _match_mailrc_line {
    my ( $line, $rules ) = @_;
    return unless defined $line;
    my ( $id,       $address ) = $line =~ m{\Aalias\s+(\S+)\s+"(.*)"};
    my ( $fullname, $email )   = $address =~ m{([^<]+)<([^>]+)>};
    $fullname =~ s/\s*$//;
    if ( $rules->{id} ) {
        return unless $rules->{id}->($id);
    }
    if ( $rules->{fullname} ) {
        return unless $rules->{fullname}->($fullname);
    }
    if ( $rules->{email} ) {
        return unless $rules->{email}->($email);
    }
    return {
        id       => $id,
        fullname => $fullname,
        email    => $email,
    };
}

1;

=for Pod::Coverage attributes validate_attributes search_packages search_authors
cached_package cached_mailrc BUILD

=head1 SYNOPSIS

  use CPAN::Common::Index::Mirror;

  # default mirror is http://www.cpan.org/
  $index = CPAN::Common::Index::Mirror->new;

  # custom mirror
  $index = CPAN::Common::Index::Mirror->new(
    { mirror => "http://cpan.cpantesters.org" }
  );

=head1 DESCRIPTION

This module implements a CPAN::Common::Index that retrieves and searches
02packages.details.txt and 01mailrc.txt indices.

The default mirror is L<http://www.cpan.org/>.  This is a globally balanced
fast mirror and is a great choice if you don't have a local fast mirror.

=cut

# vim: ts=4 sts=4 sw=4 et:


lib/CPAN/Common/Index/Mux/Ordered.pm
use 5.008001;
use strict;
use warnings;

package CPAN::Common::Index::Mux::Ordered;
# ABSTRACT: Consult indices in order and return the first result

our $VERSION = '0.011';

use parent 'CPAN::Common::Index';

use Class::Tiny qw/resolvers/;

use Module::Load ();

=attr resolvers

    An array reference of CPAN::Common::Index::* objects

=cut

sub BUILD {
    my $self = shift;

    my $resolvers = $self->resolvers;
    $resolvers = [] unless defined $resolvers;
    if ( ref $resolvers ne 'ARRAY' ) {
        Carp::croak("The 'resolvers' argument must be an array reference");
    }
    for my $r (@$resolvers) {
        if ( !eval { $r->isa("CPAN::Common::Index") } ) {
            Carp::croak("Resolver '$r' is not a CPAN::Common::Index object");
        }
    }
    $self->resolvers($resolvers);

    return;
}

=method assemble

    $index = CPAN::Common::Index::Mux::Ordered->assemble(
        MetaDB => {},
        Mirror => { mirror => "http://www.cpan.org" },
    );

This class method provides a shorthand for constructing a multiplexer.
The arguments must be pairs of subclass suffixes and arguments.  For
example, "MetaDB" means to use "CPAN::Common::Index::MetaDB".  Empty
arguments must be given as an empty hash reference.

=cut

sub assemble {
    my ( $class, @backends ) = @_;

    my @resolvers;

    while (@backends) {
        my ( $subclass, $config ) = splice @backends, 0, 2;
        my $full_class = "CPAN::Common::Index::${subclass}";
        eval { Module::Load::load($full_class); 1 }
          or Carp::croak($@);
        my $object = $full_class->new($config);
        push @resolvers, $object;
    }

    return $class->new( { resolvers => \@resolvers } );
}

sub validate_attributes {
    my ($self) = @_;
    my $resolvers = $self->resolvers;
    return 1;
}

# have to think carefully about the sematics of regex search when indices
# are stacked; only one result for any given package (or package/version)
sub search_packages {
    my ( $self, $args ) = @_;
    Carp::croak("Argument to search_packages must be hash reference")
      unless ref $args eq 'HASH';
    my @found;
    if ( $args->{name} and ref $args->{name} eq '' ) {
        # looking for exact match, so we just want the first hit
        for my $source ( @{ $self->resolvers } ) {
            if ( my @result = $source->search_packages($args) ) {
                # XXX double check against remaining $args
                push @found, @result;
                last;
            }
        }
    }
    else {
        # accumulate results from all resolvers
        my %seen;
        for my $source ( @{ $self->resolvers } ) {
            my @result = $source->search_packages($args);
            push @found, grep { !$seen{ $_->{package} }++ } @result;
        }
    }
    return wantarray ? @found : $found[0];
}

# have to think carefully about the sematics of regex search when indices
# are stacked; only one result for any given package (or package/version)
sub search_authors {
    my ( $self, $args ) = @_;
    Carp::croak("Argument to search_authors must be hash reference")
      unless ref $args eq 'HASH';
    my @found;
    if ( $args->{name} and ref $args->{name} eq '' ) {
        # looking for exact match, so we just want the first hit
        for my $source ( @{ $self->resolvers } ) {
            if ( my @result = $source->search_authors($args) ) {
                # XXX double check against remaining $args
                push @found, @result;
                last;
            }
        }
    }
    else {
        # accumulate results from all resolvers
        my %seen;
        for my $source ( @{ $self->resolvers } ) {
            my @result = $source->search_authors($args);
            push @found, grep { !$seen{ $_->{package} }++ } @result;
        }
    }
    return wantarray ? @found : $found[0];
}

1;

=for Pod::Coverage attributes validate_attributes search_packages search_authors BUILD

=head1 SYNOPSIS

    use CPAN::Common::Index::Mux::Ordered;
    use Data::Dumper;

    $index = CPAN::Common::Index::Mux::Ordered->assemble(
        MetaDB => {},
        Mirror => { mirror => "http://cpan.cpantesters.org" },
    );

=head1 DESCRIPTION

This module multiplexes multiple CPAN::Common::Index objects, returning
results in order.

For exact match queries, the first result is returned. For search queries,
results from each index object are concatenated.

=cut

# vim: ts=4 sts=4 sw=4 et:


Makefile.PL
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker 6.17;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Common library for searching CPAN modules, authors and distributions",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "CPAN-Common-Index",
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.008001",
  "NAME" => "CPAN::Common::Index",
  "PREREQ_PM" => {
    "CPAN::DistnameInfo" => 0,
    "CPAN::Meta::YAML" => 0,
    "Carp" => 0,
    "Class::Tiny" => 0,
    "File::Basename" => 0,
    "File::Copy" => 0,
    "File::Fetch" => 0,
    "File::Spec" => 0,
    "File::Temp" => "0.19",
    "File::stat" => 0,
    "HTTP::Tiny" => 0,
    "Module::Load" => 0,
    "Search::Dict" => "1.07",
    "Tie::Handle::SkipHeader" => 0,
    "URI" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Cwd" => 0,
    "Exporter" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "File::Spec::Functions" => 0,
    "Test::Deep" => 0,
    "Test::FailWarnings" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.96",
    "lib" => 0
  },
  "VERSION" => "0.011",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "CPAN::DistnameInfo" => 0,
  "CPAN::Meta::YAML" => 0,
  "Carp" => 0,
  "Class::Tiny" => 0,
  "Cwd" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Basename" => 0,
  "File::Copy" => 0,
  "File::Fetch" => 0,
  "File::Spec" => 0,
  "File::Spec::Functions" => 0,
  "File::Temp" => "0.19",
  "File::stat" => 0,
  "HTTP::Tiny" => 0,
  "Module::Load" => 0,
  "Search::Dict" => "1.07",
  "Test::Deep" => 0,
  "Test::FailWarnings" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "0.96",
  "Tie::Handle::SkipHeader" => 0,
  "URI" => 0,
  "lib" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
